cpp/ql/src/Likely Bugs/Likely Typos/illDefinedForLoop.ql

```diff
@@ -0,0 +1,102 @@
+/**
+ * @name Ill-defined for loop
+ * @description A for-loop iteration expressions goes backward with respect of the initialization statement and condition expression.
+ * @id cpp/ill-defined-for-loop
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @tags external/microsoft/6293
+ * @msrc.severity important
+ */
+import cpp
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+
+predicate illDefinedDecrForStmt( ForStmt forstmt, Variable v, Expr initialCondition, Expr terminalCondition ) { 
+  v.getAnAssignedValue() = initialCondition 
+  and
+  exists( 
+    RelationalOperation rel, Expr e |
+    rel = forstmt.getCondition() |
+    e = rel.getGreaterOperand()
+    and v.getAnAccess() = rel.getLesserOperand()
+    and terminalCondition = e
+  )
+  and
```

Comment:
Does the existing PREfast rule handle update expressions like `i += 2` or `i = i + 2`? If so, this replacement should probably do the same. If not, I wouldn't worry about it.

Reply:
No, it doesn't. I already checked that.
Do you want me to add it as a test case?

Reply:
Double-checked that PREfast will not find these cases.